### PR TITLE
feat: added a new Helm option ignoreMissingValueFiles (#7767)

### LINKED ISSUE #7767: IgnoreMissingValueFiles option to prevent passing non-existing valueFiles to helm template
# Summary

This is a proposal to add a new application's `source.helm.ignoreMissingValueFiles` property so that when set to `true`, ArgoCD would not pass files in `valueFiles` to `helm template` if they don't exist in the source under the specified path.

# Motivation

We currently use the app-of-apps pattern with a root Helm app whose application `source.helm.valueFiles` lists a static `values.yaml` and a dynamic `[cluster-fqdn]-values.yaml` file, which currently needs to exist even if empty and is used for overrides only. Needless to say, with each new cluster we end up having an additional override values file (most of which are empty).
The thing is, we also follow the same pattern for applications, and though some configurations live in a central `applications` folder at our app-of-apps repository (from the root, values are found under `applications/[app-name]/values/`) others hold that in their own repository (under `k8s/values/` - their helm is found under `k8s/[service-name]`).
This makes it very difficult to default a list of `valueFiles` following the same pattern for all our services without having to create all of these additional empty files in multiple locations.

I reckon there are ways around this to make the pattern work, but they all seem to require additional configuration or the empty files we're currently using. Inspired by a proposal mentioned by @m-yosefpor [at the end of this message](https://github.com/argoproj/argo-cd/issues/2789#issuecomment-966412660) (yet having a slightly different idea about its necessity), and realizing I'm not alone in having this question or trying to define a defaults/overrides pattern when using the app-of-apps (see the linked message and the original question it responds to), I thought to bring this up for discussion, as IMHO the best way to enable this default/overrides pattern at scale is to have a config option to purposely ignore missing values files before they're passed to helm (and fail). 

# Proposal

To be able (declaratively or via cli) to set a new Boolean `source.helm` property `ignoreMissingValueFiles` so that, when `false` (default), we'd have the current behavior (valueFiles passed to `helm template`regardless of existence), but when set to `true`, the reposerver repository `helmTemplate` processing of valuesFiles would be changed so that, for each values file in valueFiles, it would simply enable checking that, when the reference is a file, that file exists, and when it does not, it would log the omission and prevent appending it to the list of value files options that are passed to the `helm template` command. This way, configuring potentially non-existing files would not result in a failed `helm template` command.

## Note

I'm willing to submit this change myself, since from a look at the code it seems simple enough, but would appreciate feedback, recommendations, ideas and the like before I do.

### PR DESCRIPTION
**Pull request by ocraviotto**
so as to allow operators to prevent Argo CD from passing valueFiles
to helm template if they don't exist in the source under the specified path.

Signed-off-by: Oscar Craviotto <craviotto@avellaneda.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 

Closes #7767

### CODE REVIEW

File: reposerver/repository/repository.go [hunk 1]
```diff
@@ -605,6 +606,14 @@ func helmTemplate(appPath string, repoRoot string, env *v1alpha1.Env, q *apiclie
 				if err != nil {
 					return nil, err
 				}
+
+				_, err = os.Stat(path)
+				if os.IsNotExist(err) {
+					log.Warnf(" %s values file does not exist", path)
```
Comment (alexmt) [inline reposerver/repository/repository.go:612]: Most of the warning messages in logs indicate a problem that Argo CD operators have to act on. I think a warning about a missing values file is intended for the end-user only. This warning message will be printed in logs pretty frequently and won't be very useful for operators. 

I would suggest changing the log level to DEBUG. Additionally, we can introduce a "warning" condition instead. (similar to https://github.com/argoproj/argo-cd/blob/cb1f06c72ad4e52b5a8df4d016010d738b3cb1c4/pkg/apis/application/v1alpha1/types.go#L997). In my opinion "warning" condition is a nice to have a feature and I would work on it in a separate PR.

**Review by alexmt — CHANGES_REQUESTED** [review]
Thank you for a useful feature @ocraviotto ! PR looks very solid. Noticed only one minor concern about the new warning message in logs. Let me know what you think please

**Comment by alexmt** [comment]
Failed `Lint docs` CI task is definitely not related to changes in the PR. I think issue is fixed already, can you try to merge changes from "master" branch?

File: reposerver/repository/repository.go (hunk 1 above)
Comment (ocraviotto) [reply reposerver/repository/repository.go:612]: Thank you @alexmt for taking the time to start a review and for your comments. 

I agree with the above on changing the log level as this definitely affects the end-user and not the operator. TBH I did think initially to emit an event for the application resource, but your pointing me to the status conditions as an alternative to warn the user makes sense too and I'll explore it (though might leave it for a separate PR depending on how much work it'd represent and other considerations, like the proper warning condition etc.). 
In any case, my first idea was to let the end-user know about missing files, yet when looking at where in the RepoServer the file check takes place and what the various methods invoked return, I opted for the simpler log approach without much thought on how this would affect logs for operators with multiple managed resources. 

To sum up, I'll update my branch, modify the log level *and* the location of the log within the affirmative conditional below (in normal conditions the helm operation will fail and we should have information on the missing file(s) being the cause) and explore ways to update resource status conditions and at least leave a comment here on that.

Comment (ocraviotto) [reply reposerver/repository/repository.go:612]: @alexmt after a quick look, it seems to me that the best way to pass back this type of info from the RepoServer would be to add a new field to the [proto ManifestResponse message](https://github.com/argoproj/argo-cd/blob/b77f947e1dde6b200a2044780c466b364818916d/reposerver/repository/repository.proto#L45-L54) for the caller to handle the condition/event update/generation based on that field.
If his sounds reasonable, I'd be willing to take a stab at it. In that case, let me know if you'd like me to create an enhancement first or what would be the best way to proceed.

Comment (alexmt) [reply reposerver/repository/repository.go]: I agree it makes sense to use `ManifestResponse` to pass back additional information. Thank you for resolving the comment about log level! Merging this PR.

**Review by alexmt — APPROVED** [review]
